src/NServiceBus.Core/Persistence/InMemory/SagaPersister/InMemorySagaPersister.cs

```diff
@@ -66,14 +70,19 @@ public Task Save(IContainSagaData sagaData, SagaCorrelationProperty correlationP
             var inMemSession = (InMemorySynchronizedStorageSession) session;
             inMemSession.Enlist(() =>
             {
-                if (correlationProperty != SagaCorrelationProperty.None)
+                var lockenTokenKey = $"{sagaData.GetType().FullName}.{(correlationProperty == null ? "None" : correlationProperty.Name)}.{(correlationProperty == null ? "None" : correlationProperty.Value)}";
```

Comment:
The combination of saga data fullname, correlation property name and value should be unique enough to not make unnecessary lock contentions. In case when there is no correlation property we will have a lock object per saga data fullname